File: apps/desktop/lib/widgets/nano/nano_tiles.dart
import 'package:flutter/material.dart';
import '../../design/nano_tokens.dart';
import 'nano_charts.dart';
import 'nano_primitives.dart';

/// KPI summary tile used on the dashboard (label, big value, sub, optional spark).
class NanoKpiTile extends StatelessWidget {
  final String label;
  final Widget value;
  final String sub;
  final IconData icon;
  final String? tone; // 'warn' | 'crit' | null
  final List<double>? spark;
  final Color? background;
  const NanoKpiTile({
    super.key,
    required this.label,
    required this.value,
    required this.sub,
    required this.icon,
    this.tone,
    this.spark,
    this.background,
  });

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final toneColor = tone == 'crit'
        ? t.crit
        : tone == 'warn'
            ? t.warn
            : t.accent;
    final valueColor = tone == 'crit'
        ? t.crit
        : tone == 'warn'
            ? t.warn
            : t.fg;
    return Container(
      padding: const EdgeInsets.fromLTRB(14, 12, 14, 12),
      decoration: BoxDecoration(
        color: background ?? t.card,
        borderRadius: BorderRadius.circular(t.cardRadius),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: [
          Row(
            children: [
              Icon(icon, size: 14, color: t.fg3),
              const SizedBox(width: 5),
              Text(label,
                  style: TextStyle(
                      fontSize: 12, fontWeight: FontWeight.w500, color: t.fg3)),
            ],
          ),
          const SizedBox(height: 4),
          DefaultTextStyle.merge(
            style: TextStyle(
              fontSize: 26,
              fontWeight: FontWeight.w600,
              letterSpacing: -0.3,
              height: 1.1,
              color: valueColor,
              fontFeatures: const [FontFeature.tabularFigures()],
            ),
            child: value,
          ),
          const SizedBox(height: 2),
          Row(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Expanded(
                child: Text(
                  sub,
                  style: TextStyle(
                    fontSize: 10.5,
                    color: t.fg4,
                    fontFamilyFallback: kMonoFallback,
                  ),
                  overflow: TextOverflow.ellipsis,
                ),
              ),
              if (spark != null)
                NanoSparkline(data: spark!, color: toneColor),
            ],
          ),
        ],
      ),
    );
  }
}

/// Row: leading icon + short label + meter + right-aligned value/sub.
class NanoMetricRow extends StatelessWidget {
  final IconData icon;
  final String label;
  final String value;
  final String? sub;
  final double? pct; // 0..100
  final String? tone;
  const NanoMetricRow({
    super.key,
    required this.icon,
    required this.label,
    required this.value,
    this.sub,
    this.pct,
    this.tone,
  });

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final toneColor = tone == 'crit'
        ? t.crit
        : tone == 'warn'
            ? t.warn
            : t.fg;
    return Row(
      children: [
        SizedBox(width: 16, child: Icon(icon, size: 13, color: t.fg4)),
        const SizedBox(width: 10),
        SizedBox(
          width: 28,
          child: Text(label,
              style: TextStyle(
                fontSize: 11,
                color: t.fg3,
                fontFamilyFallback: kMonoFallback,
              )),
        ),
        const SizedBox(width: 10),
        if (pct != null)
          Expanded(child: NanoMeter(value: pct! / 100, color: tone == null ? null : toneColor))
        else
          const Spacer(),
        const SizedBox(width: 10),
        ConstrainedBox(
          constraints: const BoxConstraints(minWidth: 60),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.end,
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(value,
                  style: TextStyle(
                    fontSize: 13,
                    fontWeight: FontWeight.w500,
                    color: toneColor,
                    fontFeatures: const [FontFeature.tabularFigures()],
                  )),
              if (sub != null)
                Text(sub!,
                    style: TextStyle(
                      fontSize: 10,
                      color: t.fg4,
                      fontFamilyFallback: kMonoFallback,
                    )),
            ],
          ),
        ),
      ],
    );
  }
}

File: apps/ios/NanoLink/Design/NanoTiles.swift
import SwiftUI

/// KPI summary tile used on the dashboard (label, big value, sub, optional spark).
/// Ports `NanoKpiTile`.
struct NanoKpiTile<Value: View>: View {
    let label: String
    let value: Value
    let sub: String
    let icon: String
    /// `warn` | `crit` | nil
    var tone: String? = nil
    var spark: [Double]? = nil
    var background: Color? = nil
    @Environment(\.nano) private var t

    init(label: String, sub: String, icon: String, tone: String? = nil,
         spark: [Double]? = nil, background: Color? = nil,
         @ViewBuilder value: () -> Value) {
        self.label = label
        self.sub = sub
        self.icon = icon
        self.tone = tone
        self.spark = spark
        self.background = background
        self.value = value()
    }

    var body: some View {
        let toneColor = tone == "crit" ? t.crit : (tone == "warn" ? t.warn : t.accent)
        let valueColor = tone == "crit" ? t.crit : (tone == "warn" ? t.warn : t.fg)
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 5) {
                Image(systemName: icon).font(.system(size: 14)).foregroundColor(t.fg3)
                Text(label).font(.system(size: 12, weight: .medium)).foregroundColor(t.fg3)
            }
            Spacer().frame(height: 4)
            value
                .font(.system(size: 26, weight: .semibold))
                .tracking(-0.3)
                .foregroundColor(valueColor)
                .monospacedDigit()
            Spacer().frame(height: 2)
            HStack(alignment: .bottom) {
                Text(sub)
                    .font(NanoFont.mono(10.5))
                    .foregroundColor(t.fg4)
                    .lineLimit(1)
                Spacer(minLength: 0)
                if let spark {
                    NanoSparkline(data: spark, color: toneColor)
                }
            }
        }
        .padding(EdgeInsets(top: 12, leading: 14, bottom: 12, trailing: 14))
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(background ?? t.card)
        .clipShape(RoundedRectangle(cornerRadius: t.cardRadius, style: .continuous))
    }
}

/// Row: leading icon + short label + meter + right-aligned value/sub.
/// Ports `NanoMetricRow`.
struct NanoMetricRow: View {
    let icon: String
    let label: String
    let value: String
    var sub: String? = nil
    var pct: Double? = nil
    var tone: String? = nil
    @Environment(\.nano) private var t

    var body: some View {
        let toneColor = tone == "crit" ? t.crit : (tone == "warn" ? t.warn : t.fg)
        HStack(spacing: 0) {
            Image(systemName: icon).font(.system(size: 13)).foregroundColor(t.fg4)
                .frame(width: 16, alignment: .leading)
            Spacer().frame(width: 10)
            Text(label).font(NanoFont.mono(11)).foregroundColor(t.fg3)
                .frame(width: 28, alignment: .leading)
            Spacer().frame(width: 10)
            if let pct {
                NanoMeter(value: pct / 100, color: tone == nil ? nil : toneColor)
                    .frame(maxWidth: .infinity)
            } else {
                Spacer()
            }
            Spacer().frame(width: 10)
            VStack(alignment: .trailing, spacing: 0) {
                Text(value)
                    .font(.system(size: 13, weight: .medium))
                    .foregroundColor(toneColor)
                    .monospacedDigit()
                if let sub {
                    Text(sub).font(NanoFont.mono(10)).foregroundColor(t.fg4)
                }
            }
            .frame(minWidth: 60, alignment: .trailing)
        }
    }
}
